src/extension/infer.rs

```diff
@@ -107,53 +108,65 @@ pub enum InferExtensionError {
     EdgeMismatch(#[from] ExtensionError),
 }
 
-/// A graph of metavariables which we've found equality constraints for. Edges
-/// between nodes represent equality constraints.
-struct EqGraph {
-    equalities: pg::Graph<Meta, (), petgraph::Undirected>,
+/// A graph of metavariables connected by constraints.
+/// The edges represent `Equal` constraints in the undirected graph and `Plus`
+/// constraints in the directed case.
+struct GraphContainer<Dir: EdgeType> {
+    graph: pg::Graph<Meta, (), Dir>,
     node_map: HashMap<Meta, pg::NodeIndex>,
 }
 
-impl EqGraph {
-    /// Create a new `EqGraph`
-    fn new() -> Self {
-        EqGraph {
-            equalities: pg::Graph::new_undirected(),
-            node_map: HashMap::new(),
-        }
-    }
-
+impl<T: EdgeType> GraphContainer<T> {
     /// Add a metavariable to the graph as a node and return the `NodeIndex`.
     /// If it's already there, just return the existing `NodeIndex`
     fn add_or_retrieve(&mut self, m: Meta) -> pg::NodeIndex {
         self.node_map.get(&m).cloned().unwrap_or_else(|| {
-            let ix = self.equalities.add_node(m);
+            let ix = self.graph.add_node(m);
             self.node_map.insert(m, ix);
             ix
         })
     }
 
-    /// Create an edge between two nodes on the graph, declaring that they stand
-    /// for metavariables which should be equal.
-    fn register_eq(&mut self, src: Meta, tgt: Meta) {
+    /// Create an edge between two nodes on the graph
+    fn add_edge(&mut self, src: Meta, tgt: Meta) {
         let src_ix = self.add_or_retrieve(src);
         let tgt_ix = self.add_or_retrieve(tgt);
-        self.equalities.add_edge(src_ix, tgt_ix, ());
+        self.graph.add_edge(src_ix, tgt_ix, ());
     }
 
     /// Return the connected components of the graph in terms of metavariables
     fn ccs(&self) -> Vec<Vec<Meta>> {
-        petgraph::algo::tarjan_scc(&self.equalities)
+        petgraph::algo::tarjan_scc(&self.graph)
```

Comment:
Ok, I think the naming and commenting of this function has been confusing me. Graph theory says that an undirected graph has CCs but a directed graph has SCCs. (I sort of assumed that the non-strongly CC of a directed graph was just what you got by forgetting the edge directions, but that may not be common usage.)

So, petgraph's `tarjan_scc` is misnamed for undirected graphs - it returns CCs (not SCCs, there's no such thing). But naming this function CC is misnamed in the opposite way...

Ideal might be to define `sccs` for `GraphContainer<Directed>` and `ccs` for `GraphContainer<Undirected>` - that would need a macro. Alternatively if having one name for each, I'd favour misnaming it in the same way as petgraph :-( rather than another way.... and most importantly, adjust the comment.

Reply:
Renamed